plutus-core/plutus-ir/src/PlutusIR/Analysis/Definitions.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE LambdaCase #-}
+-- | Definition analysis for Plutus IR.
+-- This just adapts term-related code from PlutusCore.Analysis.Definitions;
+-- we just re-use the typed machinery to do the hard work here.
+module PlutusIR.Analysis.Definitions
+    ( UniqueInfos
+    , termDefs
+    , runTermDefs
+    ) where
+
+import Data.Functor.Foldable
+import PlutusCore.Error
+import PlutusCore.Name
+import PlutusIR.Core
+import PlutusIR.Core.Instance.Recursive (TermF (IWrapF, LamAbsF, LetF, TyAbsF, TyInstF, VarF))
+
+import Control.Monad.State
+import Control.Monad.Writer
+
+import PlutusCore.Analysis.Definitions hiding (runTermDefs, termDefs)
+
+termDefs
+    :: (Ord ann,
+        HasUnique name TermUnique,
+        HasUnique tyname TypeUnique,
+        MonadState (UniqueInfos ann) m,
+        MonadWriter [UniqueError ann] m)
+    => Term tyname name uni fun ann
+    -> m ()
+termDefs =
+    let -- Add bindings to definition maps.
+        -- addBindingDef :: Binding tyname name uni fun ann -
+        addBindingDef (TermBind _tmAnn _tmStrict (VarDecl varAnn n varTy) tm) =
+            addDef n varAnn TermScope >> typeDefs varTy >> tm
```

Comment:
This is where recursion schemes don't work. Recursion schemes work for a self-recursive datatype. Here you have a term which is getting totally ignored because the recursion schemes don't "see" the recursion over here. You are going to need to rewrite it to not use recursion schemes: they just don't work for PIR because of the mutually recursive datatypes.

Reply:
Recursion schemes can work for mutually recursive datatypes but I also think the machinery is too heavy just for testing global uniqueness.

Reply:
Like I said elsewhere, these days I'm much more in favour of using our sub-term traversals rather than using `recursion-schemes`, I'd happily purge them from our codebase entirely at this point.

Reply:
I'll change that out to sub-term traversals like in the inliner. Since this is not a trivial function, it probably should have some tests, like [multiplyDefined](https://github.com/input-output-hk/plutus/blob/a667f429516f4246e74f175e0f9407e57a4048ca/plutus-core/plutus-core/test/Check/Spec.hs#L55) but for PIR. I'll add that. I'll re-request review when I'm ready.

Reply:
I highly recommend updating the PLC one too if  you do that. They should be defined extremely similarly, so it should be simple enough to replace the PLC one (just delete the cases for PIR terms).